Issue 1306 rename website breaks grid

phoenix128 opened the pull request:
### Fixed Issues (if relevant)
1. magento-engcom/msi#1306: Renaming website breaks MSI grid

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

phoenix128 committed:
FIX for issue https://github.com/magento-engcom/msi/issues/1306
If a website was renamed/removed, sales channels were still tying to search for it

phoenix128 committed:
Keep data consistency with inventory_stock_sales_channel when a website code is changed
See https://github.com/magento-engcom/msi/issues/1306

maghamed commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,85 @@
+        int $websiteId
+    ) {
+        $connection = $resourceModel->getConnection();
+        $qry = $connection->select()->from($resourceModel->getMainTable(), 'code')->where('website_id = ?', $websiteId);
```
please don't use shortening in variable names 
like `$qry` 
it's better to have long variable names than shortening

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,85 @@
+    private function getCodeFromDatabase(
+        \Magento\Store\Model\ResourceModel\Website $resourceModel,
+        int $websiteId
+    ) {
```
Pls, 
1) use return type hinting
2) `$qry` is not clear name for the variable
3) Need to use `$this->resource` instead of `$resourceModel->getConnection()`, 
generally we don't need dependency on `ResourceModel\Website` for obtaining website code
4) Also, need to think about extracting this logic to separate class

phoenix128 replied:
I used `\Magento\Store\Model\ResourceModel\Website` with intention because it can be potentially stored in a different database from the standard one.

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,85 @@
+        $tableName = $this->resource->getTableName('inventory_stock_sales_channel');
+        $connection->update($tableName, [
+            SalesChannelInterface::CODE => $newCode,
+        ], $connection->quoteInto(SalesChannelInterface::CODE . "=? and type='website'", $oldCode));
```
1) Use `\Magento\InventorySalesApi\Api\Data\SalesChannelInterface::TYPE_WEBSITE
2) Second parameter can be array with clauses instead of `sql string`

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,85 @@
+        if ($object->getId()) {
+            // Keep database consistency while updating the website code
+            // See https://github.com/magento-engcom/msi/issues/1306
+            $oldCode = $this->getCodeFromDatabase($subject, (int) $object->getId());
```
Bad interface of method

naydav commented on app/code/Magento/InventorySalesAdminUi/Ui/Component/Listing/Column/SalesChannels.php:
```diff
@@ -70,10 +71,16 @@ private function prepareSalesChannelData(array $salesChannelData): array
+                        'code' => $code,
+                    ];
+                } catch (NoSuchEntityException $e) {
+                    // We must handle the scenario in which a previous website was removed/renamed
```
I'm not sure that is good idea to skip exception... especially if we fixed initial case

phoenix128 replied:
I agree, but this is the only way to fix the already done mistake.

maghamed commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,85 @@
+     * @param \Magento\Store\Model\ResourceModel\Website $subject
+     * @param callable $proceed
+     * @param \Magento\Store\Model\Website $object
+     * @return mixed
```
why do you have mixed as return type? 
you can specify `\Magento\Store\Model\ResourceModel\Website` as resource model::save returns this

phoenix128 replied:
Phpstorm autocompletition... my fault I did not notice it.

phoenix128 committed:
Refactoring for AssignWebsiteToDefaultStock moving it from observer to around plugin
to keep data consistent while chaning the website code on an existing website.
This commit is realted to issue https://github.com/magento-engcom/msi/issues/1306

phoenix128 committed:
FIX PHPMD + PHPCS validation

phoenix128 commented:
@naydav @maghamed , tests are failing because `Magento_Catalog` does not have a sequence entry in `module.xml` for `Magento_Store`. It should be since tests and business logic require it.

**Magento/Catalog/etc/module.xml**:

```
<?xml version="1.0"?>
<!--
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
-->
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
    <module name="Magento_Catalog" >
        <sequence>
            <module name="Magento_Eav"/>
            <module name="Magento_Cms"/>
            <module name="Magento_Indexer"/>
            <module name="Magento_Customer"/>
        </sequence>
    </module>
</config>
```

It happens on this module because the fixture `dev/tests/integration/testsuite/Magento/Catalog/Model/ResourceModel/_files/website_attribute_sync.php` it is updating `core_website` table through resource model:

```
$website = $objectManager->create(Website::class);
$website->setName('custom website for av test')
    ->setCode('customwebsite1');
$website->isObjectNew(true);
$websiteResourceModel->save($website);
```

Should I fix this issue in the core or temporary skip this exception?

Skipping the exception in `\Magento\InventorySales\Plugin\Store\WebsiteResourcePlugin::aroundSave` is not my best option of course, but looking at current tests is not a big issue.

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,92 @@
+    private function getCodeFromDatabase(int $websiteId): string
+    {
+        $connection = $this->resourceConnection->getConnection();
+        $tableName = $connection->getTableName('store_website');
```
We need to extract table name via `$this->resourceConnection->getTableName()`
Because `$connection->getTableName` is only wrapper which doesn't take into account table prefix

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,92 @@
+        $connection = $this->resourceConnection->getConnection();
+        $tableName = $connection->getTableName('store_website');
+        $selectQry = $connection->select()->from($tableName, 'code')->where('website_id = ?', $websiteId);
+        return (string) $connection->fetchOne($selectQry);
```
I'm not sure that we need to return `''`(empty string) if the record was not found (website id is not existing)
More clear is return null in this case

phoenix128 replied:
That method can't return an empty string. If it is called, it is called from `aroundSave` and only if the website exists.

Maybe I should change the initialization of `$oldCode = '';` to `$oldCode = null;`.
And change the `if` statement accordingly:

```
...
if (($oldCode !== null) && ($oldCode !== $newCode)) {
...
```

naydav commented on app/code/Magento/InventorySales/Plugin/Store/WebsiteResourcePlugin.php:
```diff
@@ -0,0 +1,92 @@
+use Magento\InventorySales\Model\AssignWebsiteToDefaultStock;
+use Magento\InventorySales\Model\ResourceModel\UpdateSalesChannelWebsiteCode;
+
+class WebsiteResourcePlugin
```
No bad to have `what does the plugin do` in the class name

naydav commented on app/code/Magento/InventorySales/etc/events.xml:
```diff
@@ -6,9 +6,6 @@
-    <event name="website_save_after">
-        <observer name="assign_website_to_default_stock" instance="Magento\InventorySales\Observer\Website\AssignWebsiteToDefaultStock"/>
-    </event>
     <event name="website_delete_after">
```
Looks like we also need to update delete logic `website_delete_after`

phoenix128 replied:
An observer after delete seems to be an efficient way and seems to be working, do you prefer to pluginize it?

phoenix128 replied:
Done in: https://github.com/magento-engcom/msi/pull/1356

phoenix128 committed:
Modified plugin class name to reflect its responsability + switched to correct getTableName method

phoenix128 committed:
Return null from getCodeFromDatabase if the required website does not exist

phoenix128 committed:
Created interface and preference for UpdateSalesChannelWebsiteCode

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=1344) <br/>All committers have signed the CLA.

phoenix128 committed:
Integration test for website rename scenario and stock link consistency

Commit:
MSI-1306: Rename website breaks grid

Commit:
Merge remote-tracking branch 'origin/2.3-develop' into issue-1306-rename-website-breaks-grid

Commit:
MSI-1306: Rename website breaks grid

Commit:
MSI-1306: Rename website breaks grid

-- fix static test

smoskaluk commented:
@phoenix128  Problem is not fixed, please recheck. I changed code(or name) of Main Website and saved it - and 've got fatal error:
Fatal error: Uncaught Error: Call to undefined method Magento\Framework\App\ResourceConnection\Interceptor::execute() in /var/www/html/magento2ce/pr7/app/code/Magento/InventorySales/Plugin/Store/Model/ResourceModel/Website/UpdateSalesChannelWebsiteCodePlugin.php:64 Stack trace: #0 /var/www/html/magento2ce/pr7/lib/internal/Magento/Framework/Interception/Interceptor.php(135): Magento\InventorySales\Plugin\Store\Model\ResourceModel\Website\UpdateSalesChannelWebsiteCodePlugin->aroundSave(Object(Magento\Store\Model\ResourceModel\Website\Interceptor), Object(Closure), Object(Magento\Store\Model\Website\Interceptor)) #1 /var/www/html/magento2ce/pr7/lib/internal/Magento/Framework/Interception/Interceptor.php(153): Magento\Store\Model\ResourceModel\Website\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Store\Model\Website\Interceptor)) #2 /var/www/html/magento2ce/pr7/generated/code/Magento/Store/Model/ResourceModel/Website/Interceptor.php(130): Magento\Store\Model\ResourceModel\Website\Interceptor->___callPlu in /var/www/html/magento2ce/pr7/app/code/Magento/InventorySales/Plugin/Store/Model/ResourceModel/Website/UpdateSalesChannelWebsiteCodePlugin.php on line 64

phoenix128 commented:
@smoskaluk , seems it is not coming from my original code, but from other commits, let me check it.

Commit:
MSI-1306: Rename website breaks grid

naydav commented:
Sorry, It was a mistake from my side
Already fixed

Commit:
MSI-1306: Rename website breaks grid
